python/ola/PidStoreTest.py

```diff
@@ -198,6 +198,87 @@ def testCmp(self):
     self.assertNotEqual(hash(p1b), hash(p2))
     self.assertNotEqual(hash(p1a), hash(p3))
 
+  def testPackUnpack(self):
+    store = PidStore.PidStore()
+    store.Load([os.path.join(path, "test_pids.proto")])
+
+    pid = store.GetName("DMX_PERSONALITY_DESCRIPTION")
+
+    # Pid.Pack only packs requests and Pid.Unpack only unpacks responses
+    # so test in two halves
+    args = ["42"]
+    blob = pid.Pack(args, PidStore.RDM_GET)
+    decoded = pid._requests.get(PidStore.RDM_GET).Unpack(blob)[0]
+    self.assertEqual(decoded['personality'], 42)
+
+    args = ["42", "7", "UnpackTest"]
+    blob = pid._responses.get(PidStore.RDM_GET).Pack(args)[0]
+    decoded = pid.Unpack(blob, PidStore.RDM_GET)
+    self.assertEqual(decoded['personality'], 42)
+    self.assertEqual(decoded['slots_required'], 7)
+    self.assertEqual(decoded['name'], "UnpackTest")
+
+  def testPackRanges(self):
+    store = PidStore.PidStore()
+    store.Load([os.path.join(path, "test_pids.proto")])
+
+    pid = store.GetName("REAL_TIME_CLOCK")
+
+    args = ["2020", "6", "20", "20", "20", "20"]
+    blob = pid.Pack(args, PidStore.RDM_SET)
+    self.assertTrue(len(blob) > 1)
```

Comment:
Any reason not to test this blob is exactly the intended length (and possibly test a data match)?

Reply:
Doesn't hurt to check.  I was thinking of the range validation in that test and mostly threw that in to check that the valid ranges don't throw, but I'll decode it and confirm it is correct.

Reply:
Looks good thanks.

Reply:
Actually, thinking about it, comparing to a byte array might be better. If pack ended up reversing the data or doing incorrect network/host order for the larger values, this potentially wouldn't catch it, whereas confirming the packed value matches the expected would. I think we're already testing unpack elsewhere aren't we?

Reply:
RDMTest.py has the byte-level encoding validation of a pretty large structure with at least one large value.  So I think that's captured and given the problem updating all these tests if the protobuf is ever changed, IMHO it would be better not to check the wire encoding in too many places.

Reply:
pid.Pack should be the RDM wire encoding (which is then put in the Protobuf):
https://github.com/OpenLightingProject/ola/blob/0028d658d6e58ec5158c017ae676f30103d75bbc/python/ola/PidStore.py#L210-L219

So if it changes we're rewriting all of OLA's RDM implementation!

See the C++ equivalent here for example: https://github.com/OpenLightingProject/ola/blob/0028d658d6e58ec5158c017ae676f30103d75bbc/common/rdm/MessageDeserializerTest.cpp

Reply:
My objection is not so much how or whether it might change, but that I think all of the wire/encoding tests of outbound get and set requests with and without parameters as well as inbound responses with data and nacks are fairly thoroughly testing in RDMTest.py, and adding them additionally here where the intention was to just check that the ranges are being enforced correctly isn't necessary and only confuses things.  But it doesn't hurt a whole lot and since I wanted to push the other change, I'll check the blob.

Reply:
Ah okay, I assumed you were testing the RDM serialisation in PidStore and the more general Protobuf stuff in RDMTest. Thanks for adding it anyway.